docs/adr/0006-consensus-governance.md

```diff
@@ -0,0 +1,478 @@
+
+- **List of votes (`0x83`)**
+
+  Each vote is stored under a separate storage key with the following key
+  format:
+
+  ```
+  0x83 <proposal-id (uint64)> <voter-address (staking.Address)>
+  ```
+
+  And CBOR-serialized value:
+
+  ```golang
+  // Vote is a governance vote.
+  type Vote uint8
+
+  const (
+      VoteYes     Vote = 1
+      VoteNo      Vote = 2
+      VoteAbstain Vote = 3
+  )
+  ```
+
+- **List of pending upgrades (`0x84`)**
+
+  Each pending upgrade is stored under a separate storage key with the following
+  key format:
+
+  ```
+  0x84 <upgrade-epoch (beacon.EpochTime)> <proposal-id (uint64)>
+  ```
+
+### Genesis Document
+
+The genesis document needs to be updated to include a `governance` field with
+any initial state (see [_State_]) and consensus parameters (see [_Consensus
+Parameters_]) for the governance service.
+
+[_State_]: #state
+[_Consensus Parameters_]: #consensus-parameters
+
+### Transaction Methods
+
+This proposal adds the following transaction methods in the governance module:
+
+#### Submit Proposal
+
+Proposal submission enables a new consensus layer governance proposal to be
+created.
+
+**Method name:**
+
+```
+gov.SubmitProposal
+```
+
+**Body:**
+
+```golang
+// ProposalContent is a consensus layer governance proposal content.
+type ProposalContent struct {
+    Upgrade       *UpgradeProposal       `json:"upgrade,omitempty"`
+    CancelUpgrade *CancelUpgradeProposal `json:"cancel_upgrade,omitempty"`
+}
+
+// UpgradeProposal is an upgrade proposal.
+type UpgradeProposal struct {
+    upgrade.Descriptor
+}
+
+// CancelUpgradeProposal is an upgrade cancellation proposal.
+type CancelUpgradeProposal struct {
+    // ProposalID is the identifier of the pending upgrade proposal.
+    ProposalID uint64 `json:"proposal_id"`
+}
+```
+
+**Fields:**
+
+- `upgrade` (optional) specifies an upgrade proposal.
+- `cancel_upgrade` (optional) specifies an upgrade cancellation proposal.
+
+Exactly one of the proposal kind fields needs to be non-nil, otherwise the
+proposal is considered malformed.
+
+Upon processing any proposal the following steps are first performed:
+
+- The account indicated by the signer is loaded.
+
+- If the account balance is less than `min_proposal_deposit`, the method call
+  fails with `ErrInsufficientBalance`.
+
+Upon processing an **`UpgradeProposal`** the following steps are then performed:
+
+- The upgrade descriptor is checked for basic internal validity. If the check
+  fails, the method call fails with `ErrInvalidArgument`.
+
+- The upgrade descriptor's `epoch` field is compared with the current epoch. If
+  the specified epoch is not at least `upgrade_min_epochs_diff` epochs ahead of
+  the current epoch, the method call fails with `ErrUpgradeTooSoon`.
+
+- The set of pending upgrades is checked to make sure that no upgrades are
+  currently pending. If there is an existing upgrade pending, the method call
+  fails with `ErrUpgradeAlreadyPending`.
```

Comment:
This seems a little odd:  we prevent new proposals when one is pending, but as long as active proposals have not been voted on and moved to pending state, additional proposals can be made.  Is the intent to prevent two pending proposals from becoming ... "in use" (i.e., approved and switch-over epoch has passed, so presumably all node operators have picked up the implementation or implemented the proposal themselves)?  What happens (if anything) if two or more active proposals go into pending state and their switch-over epochs are close to each other, e.g., within upgrade_min_epochs_diff of each other?

Not a critical issue, but just wanted clarification / trying to understand the intent.

Reply:
> we prevent new proposals when one is pending, but as long as active proposals have not been voted on and moved to pending state, additional proposals can be made

I think this is just a misunderstanding, there are two things to an upgrade:
* First a governance upgrade proposal is created and it goes through the voting process.
* Only if the vote passes is an upgrade pending.

So this does not refer to other pending governance upgrade proposals but to already passed pending upgrades.

> What happens (if anything) if two or more active proposals go into pending state and their switch-over epochs are close to each other, e.g., within upgrade_min_epochs_diff of each other?

See the _Proposal Content Execution_ section below where there is another check when the passed governance upgrade proposal is executed. Only one upgrade can be pending and if two upgrade proposals are passed, the second one will fail to execute.